core/os/device/ggp/device.go

```diff
@@ -0,0 +1,285 @@
+			filepath.Join(ggpSDKPath, "dev", "bin", ggpExe))
+	}
+	for _, path := range search {
+		if p, err := file.FindExecutable(path); err == nil {
+			GGP = p
+			return GGP, nil
+		}
+	}
+
+	return file.Path{}, fmt.Errorf(
+		"ggp could not be found from GGP_SDK_PATH or PATH\n"+
+			"GGP_SDK_PATH: %v\n"+
+			"PATH: %v\n"+
+			"search: %v",
+		os.Getenv("GGP_SDK_PATH"), os.Getenv("PATH"), search)
+}
+
+func ggpDefaultRootConfigPath() (string, error) {
+	if runtime.GOOS == "windows" {
+		return os.Getenv("APPDATA"), nil
+	}
+	if p := os.Getenv("XDG_CONFIG_HOME"); p != "" {
+		return path.Clean(p), nil
+	}
+	if p := os.Getenv("HOME"); p != "" {
+		return path.Join(path.Clean(p), ".config"), nil
+	}
+	return "", fmt.Errorf("Can not find environment")
+}
+
+type GGPConfiguration struct {
+	remotessh.Configuration
+	Gamelet string
+}
+
+func getConfigs(ctx context.Context) ([]GGPConfiguration, error) {
+	configs := []GGPConfiguration{}
+
+	ggpPath, err := GGPExecutablePath()
+	if err != nil {
+		return nil, log.Errf(ctx, err, "Could not find ggp executable to list gamelets")
+	}
+	cli := ggpPath.System()
+
+	cmd := shell.Command(cli, "gamelet", "list")
+	gameletListOutBuf := &bytes.Buffer{}
+	gameletListErrBuf := &bytes.Buffer{}
+
+	if err := cmd.Capture(gameletListOutBuf, gameletListErrBuf).Run(ctx); err != nil {
+		return nil, err
+	}
+
+	t, err := ParseListOutput(gameletListOutBuf)
+	if err != nil {
+		return nil, log.Errf(ctx, err, "parse gamelet list")
+	}
+	for _, inf := range t.Rows {
+		if inf[3] != "RESERVED" && inf[3] != "IN_USE" {
+			continue
+		}
+		sshInitOutBuf := &bytes.Buffer{}
+		sshInitErrBuf := &bytes.Buffer{}
+
+		if err := shell.Command(cli, "ssh", "init", "--gamelet", inf[1], "-s").Capture(sshInitOutBuf, sshInitErrBuf).Run(ctx); err != nil {
+			log.W(ctx, "'ggp ssh init --gamelet %v -s' finished with error: %v", inf[1], err)
+			continue
+		}
+		if sshInitErrBuf.Len() != 0 {
+			log.W(ctx, "'ggp ssh init --gamelet %v -s' finished with error: %v", inf[1], sshInitErrBuf.String())
+			continue
+		}
+		envs := []string{
+			"YETI_DISABLE_GUEST_ORC=1",
+			"YETI_DISABLE_STREAMER=1",
+		}
+		cfg := remotessh.Configuration{
+			Name: inf[0],
+			Env:  envs,
+		}
+		if err := json.Unmarshal(sshInitOutBuf.Bytes(), &cfg); err != nil {
+			log.W(ctx, "Failed at unmarshaling 'ggp ssh init --gamelet %v -s' output, fallback to use default config, err: %v", inf[1], err)
+		}
+		configs = append(configs, GGPConfiguration{cfg, inf[1]})
+	}
+	return configs, nil
+}
+
+// Monitor updates the registry with devices that are added and removed at the
+// specified interval. Monitor returns once the context is cancelled.
+func Monitor(ctx context.Context, r *bind.Registry, interval time.Duration) error {
+	unlisten := registry.Listen(bind.NewDeviceListener(r.AddDevice, r.RemoveDevice))
+	defer unlisten()
+
+	for _, d := range registry.Devices() {
+		r.AddDevice(ctx, d)
+	}
+
+	var lastErrorPrinted time.Time
+
+	for {
+		if err := func() error {
```

Comment:
This whole closure looks pointless.

Isn't it the same as:

```go
		configs, err := getConfigs(ctx)
		if err != nil {
			return err
		}
		if err := scanDevices(ctx, configs); err != nil {
			if time.Since(lastErrorPrinted).Seconds() > printScanErrorsEveryNSeconds {
				log.E(ctx, "Couldn't scan devices: %v", err)
				lastErrorPrinted = time.Now()
			}
		} else {
			lastErrorPrinted = time.Time{}
		}
```

Reply:
Yes! Done.